feat: add legacy completions API support

roman-romanov-o opened the pull request:
Closes https://github.com/epam/ai-dial-adapter-openai/issues/112

Commit:
refactor: move env constants in distinct file, move exception handler in utils

Commit:
add string processors abstraction

Commit:
add legacy completions endpoint, not completed testing

ai-dial-actions commented:
Hey there and thank you for opening this pull request! 👋🏼

We require pull request titles to follow the [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/) and it looks like your proposed title needs to be adjusted.

Details:

```
Unknown release type "Feature" found in pull request title "Feature: add legacy completions API support". 

Available types:
 - fix
 - hotfix
 - feat
 - feature
 - chore
```
<!-- Sticky Pull Request Commentpr-title-lint-error -->

Commit:
- remove string processor for simpler version
- patch processor of responses to accept streams without header in response

Commit:
enchance patching for cases, when stream response is without text/event-stream header

roman-romanov-o commented:
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-openai-pr-113.nightly-test.deltixhub.io
>E2E tests status: failed

adubovik commented on aidial_adapter_openai/legacy_completions.py:
```diff
@@ -0,0 +1,102 @@
+
+
+async def legacy_completions(
+    data,
```
Let's add type hints to the arguments

roman-romanov-o replied:
fixed, but adding type hints to data cause adding typing ignore somewhere, because linter thinks that

`data.get('n', 1)` still can be None

adubovik replied:
`data: dict` is enough :)

I want all of the parameters to be typed to a reasonable extent, which is the case for the rest of the project.

roman-romanov-o replied:
I think, that such `data: dict` type hints are mostly useless, other than preventing some really basic errors.
It's the same as adding `data: object` or `data: Any`.

Even when I know that data is `dict/list/object`, I have no idea at all, how to actually work with this parameter inside of code.

roman-romanov-o replied:
I understand that it's hard to add specific types to all nested objects, but at least we can do it in some places, to make life of developer easier, and prevent him from making `print(data)` to actually understand, what's hidden inside of this dictionary

adubovik commented on aidial_adapter_openai/legacy_completions.py:
```diff
@@ -0,0 +1,102 @@
+            {
+                "index": choice["index"],
+                "finish_reason": choice["finish_reason"],
+                "message": {
```
In the streaming mode the `message` field is called `delta`.

roman-romanov-o replied:
fixed

adubovik commented on aidial_adapter_openai/app.py:
```diff
@@ -43,50 +51,10 @@
 init_telemetry(app, TelemetryConfig())
 configure_loggers()
 
-model_aliases: Dict[str, str] = json.loads(os.getenv("MODEL_ALIASES", "{}"))
```
Let's make the refactoring in a dedicated MR.

adubovik commented on aidial_adapter_openai/app.py:
```diff
@@ -100,30 +68,40 @@ def get_api_version(request: Request):
     is_stream = data.get("stream", False)
 
-    api_type, api_key = await get_credentials(request, chat_completions_parser)
+    if deployment_id in LEGACY_COMPLETIONS_DEPLOYMENTS:
```
You really don't need to introduce `LEGACY_COMPLETIONS_DEPLOYMENTS` env var, because the format of the upstream endpoint already tells you if this is a chat completions model of a legacy completions model. The parsers do the work of distinguishing one from another.

roman-romanov-o replied:
fixed

adubovik commented on aidial_adapter_openai/openai_override.py:
```diff
@@ -31,11 +35,90 @@ def handle_error_response_wrapper(wrapped, self, args, kwargs):
     raise OpenAIException(*args)
 
 
+# Overriden to handle cases, when server returns valid stream, but wihtout "text/event-stream" header
```
wihtout => without

roman-romanov-o replied:
fixed

Commit:
- added a bit more typing
- remove legacy deployments from env vars
- detect legacy completions requests by endpoint url
- minor fixes

Commit:
Minor fixes

roman-romanov-o commented:
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-openai-pr-113.nightly-test.deltixhub.io
>E2E tests status: success

Commit:
remove extra params for completions, if they are not in original data

Commit:
minor fix - add possibility to pass engine as param

Commit:
allow empty usage inside of response

adubovik commented on aidial_adapter_openai/legacy_completions.py:
```diff
@@ -0,0 +1,117 @@
+
+
+def convert_completion_to_chat_completion(
+    item: OpenAIObject, role="assistant", is_stream: bool = False
```
The defaults here do not make much sense. Why non-streaming is more preferable default than streaming? They are both on the equal footing to me.
Same with the role parameter.

roman-romanov-o replied:
fixed, removed defaults

adubovik commented on aidial_adapter_openai/legacy_completions.py:
```diff
@@ -0,0 +1,117 @@
+            {
+                "index": choice["index"],
+                "finish_reason": choice["finish_reason"],
+                ("message" if is_stream else "delta"): {
```
It's other way around. Consider reusing the helper:

https://github.com/epam/ai-dial-adapter-openai/blob/ce4a06b308123fa5d3d07ccbc87be4ee252d349d/aidial_adapter_openai/utils/streaming.py#L23-L45

adubovik commented on aidial_adapter_openai/legacy_completions.py:
```diff
@@ -0,0 +1,117 @@
+        "choices": [
+            {
+                "index": choice["index"],
+                "finish_reason": choice["finish_reason"],
```
Are you sure "finish_reason" will be present in every single chunk of response? Why not `choice.get("finish_reason")`?

adubovik commented on aidial_adapter_openai/utils/remove_none.py:
```diff
@@ -0,0 +1,11 @@
+
+
+def remove_none(d: Dict[str, Optional[Any]]) -> Dict[str, Any]:
+    """
```
IMO, 6 lines of comments for 1 line of actual code is an overkill.

Noone ever is going to read this comment from start to finish. 
One would jump straight away to the line of code.

roman-romanov-o replied:
Currently it doesn't make sense, you're right

But it could make sense to add such `reStructuredText` (or maybe google/numpy type of comments) to each function, so at some point we could generate automatic documentation using some tool like `Sphinx`

adubovik replied:
I agree, but we don't have such plans yet. 

Also, automatic documentation is a good idea for libraries, but this adapter isn't a library but an application on its own.

adubovik commented on aidial_adapter_openai/legacy_completions.py:
```diff
@@ -0,0 +1,117 @@
+    api_type: str,
+    api_version: str,
+    api_base: str,
+    model=None,
```
```
model: str | None = None,
engine: str | None = None
```

adubovik commented on aidial_adapter_openai/utils/parsers.py:
```diff
@@ -53,22 +58,32 @@ def parse(self, endpoint: str) -> AzureOpenAIEndpoint | OpenAIEndpoint:
+        return match
+
+    def is_valid(self, endpoint: str) -> bool:
+        if self._find_match(endpoint):
```
`return self._find_match(endpoint)`

adubovik commented on aidial_adapter_openai/utils/parsers.py:
```diff
@@ -53,22 +58,32 @@ def parse(self, endpoint: str) -> AzureOpenAIEndpoint | OpenAIEndpoint:
+        return False
+
+    def parse(self, endpoint: str) -> AzureOpenAIEndpoint | OpenAIEndpoint:
+        if match := self._find_match(endpoint):
```
`return self._find_match(endpoint)`

roman-romanov-o replied:
`self._find_match(endpoint)`could be None, need to check it before returning

adubovik commented on aidial_adapter_openai/utils/parsers.py:
```diff
@@ -53,22 +58,32 @@ def parse(self, endpoint: str) -> AzureOpenAIEndpoint | OpenAIEndpoint:
+        return match
+
+    def is_valid(self, endpoint: str) -> bool:
+        if self._find_match(endpoint):
```
`return self._find_match(endpoint) is not None`

roman-romanov-o replied:
Done

adubovik commented on aidial_adapter_openai/app.py:
```diff
@@ -100,15 +67,37 @@ def get_api_version(request: Request):
 
-    api_type, api_key = await get_credentials(request, chat_completions_parser)
+    # If it is not chat completions, but legacy completions
+    if not chat_completions_parser.is_valid(
```
It's better to parse and validate at the same time, so that you don't need to keep in mind that every `parse` should be only called under the branch where `is_valid` is True:

```python
if (endpoint := completions_parser.parse(upstream_endpoint)) is not None:
    ...
```

roman-romanov-o replied:
This would not work with current parsers, since they are raising `DialException` if parsing has not succeeded
I don't really want to change behaviour of this class 

---

Also, another problem, that raises when we send both `/chat/completions` and `/completions` to parser, is that with current `EndpointParser`,

`EndpointParser(name="completions").parse("/some_deployment/chat/completions") will parse successfully, with
`deployment_id = some_deployment/chat`

roman-romanov-o replied:
The approach here, other than I use right now (check that chat completions is not valid in same statement, to filter chat completions from completions) is to create subclass

```python
class CompletionsEndpointParser(EndpointParser):
    ...
```
to exclude `chat/completions` case

adubovik replied:
> This would not work with current parsers

The parsers aren't set in stone.

This condition looks so unnatural, because we didn't bother to fix the parsers when `/completions` case was introduced. Conceptually parsers should be disjoint.

Commit:
refactor parsers

Commit:
remove default parameters from convertion of completion to chat completion, better typings
